### PR TITLE
Add "set as default" checkbox to download page for Windows (#16313)

### LINKED ISSUE #16313: Add "set as default browser" checkbox to download page for Windows
## Description

We're promoting the experiment in https://github.com/mozilla/bedrock/issues/15895 into production.

See #16020 

---

#### Success Criteria
- [ ] Windows users see a checkbox to set Firefox as their default browser when installing

### PR DESCRIPTION
**Pull request by stephaniehobson**
## One-line summary

Add "set as default" checkbox to download page for Windows (#16313)

## Significant changes and points to review

- Adds checkbox to both refresh and un-refreshed download pages
   - If the switch is enabled
   - For Windows users with a version higher than Windows 8.1
   - If the checkbox string is translated
   - If they are not in a country that requires consent to store data
- Adds unit/jasmine tests for new JS
- Does not add functional/playwright tests for page changes
  - We can't mock the OS with playwright
- Improves documentation for download pages

## Issue / Bugzilla link

Fix #16313

## Testing

It looks like the entire file /firefox-new-refresh.html changed because I fixed the indentation, turn off white space to see a better diff.

In Windows:

- [ ] (checkbox) http://localhost:8000/en-US/firefox/new/ 
- [ ] (no checkbox) http://localhost:8000/en-US/firefox/new/?geo=FR
- [ ] (checkbox - but it needs the string so you have to fake that) http://localhost:8000/fr/firefox/new/
- [ ] (no checkbox) http://localhost:8000/fr/firefox/new/?geo=FR

In MacOS/Linux/iOS etc:

- [ ] (no checkbox) http://localhost:8000/en-US/firefox/new/ 
- [ ] (no checkbox) http://localhost:8000/en-US/firefox/new/?geo=FR
- [ ] (no checkbox - but it needs the string so you have to fake that) http://localhost:8000/fr/firefox/new/
- [ ] (no checkbox) http://localhost:8000/fr/firefox/new/?geo=FR

### CODE REVIEW

File: docs/docs/firefox-downloads/download-as-default.md [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+- reveal the checkboxes
+- strip utm parameters from the URL
+- update the URL to include the `SET_DEFAULT_BROWSER` campaign parameter
+- refresh stub attribution to reflect the updated url
+
+Unchecking the checkbox will not revert to the previous value of the stub attribution code.
+
+Note that it is permitted by EU/EAA law to send the campaign parameter in response to an explicit
+user action so we may enable this for `needs_data_consent` countries in the future. One blocker
```
Comment (maureenlholland) [inline docs/docs/firefox-downloads/download-as-default.md]: **docs, non-blocking**: remove unfinished content for now, we can add later

```suggestion
user action so we may enable this for `needs_data_consent` countries in the future. 
```

File: docs/docs/firefox-downloads/download-as-default.md [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+is that
```
Comment (maureenlholland) [inline docs/docs/firefox-downloads/download-as-default.md]:
```suggestion
```

**Review by maureenlholland — COMMENTED** [review]
Code looks good, moving on to browser testing

My only thought on testing is whether it's worth reinforcing the order of the meetsRequirement IF statements with specific test cases. (i.e. User in EU with consent cookie analytics true returns true)

File: media/js/firefox/new/desktop/download-as-default.es6.js [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+    return true;
+};
+
+/**
+ * Init Firefox set-as-default opt-out flow.
+ * @returns {Boolean}.
+ */
+DownloadAsDefault.init = () => {
+    if (!DownloadAsDefault.meetsRequirements()) {
+        return false;
```
Comment (maureenlholland) [inline media/js/firefox/new/desktop/download-as-default.es6.js:227]: **issue (potentially blocking, but could also be a fast-follow for an edge case)**: while browser testing, I forgot to clear the stub attribution cookie from a non-EU location (checked checkbox by default). When I then went to test an EU location, I did not get the checkbox, but I still had the set as default stub attribution code, with no way to remove it.

I see two options:
1. Always show checkboxes if there is an existing "set default browser" stub attribution code in cookies. Gives user option to change their mind (but might introduce other problems and definitely more complicated because we'd have to decode the existing attribution code to check if it has this campaign)
2. Re-initiate stub attribution if current context doesn't meet download as default requirements to override the "set default browser" stub attribution.

File: media/js/firefox/new/desktop/download-as-default.es6.js [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+    } else {
+        url = DownloadAsDefault.addUTMParams(url);
+    }
+
+    // Update the browser's URL without reloading
+    window.history.replaceState({}, document.title, url);
+
+    /**
+     * Finally, re-initiate attribution based on the new URL
+     * parameters. Only rebind events after attribution
```
Comment (maureenlholland) [inline media/js/firefox/new/desktop/download-as-default.es6.js:102]: I tried to test a case where the user quickly clicks the checkbox and the Download button (before the checkbox has re-enabled) to see if it was possible to get an outdated stub attribution code in the download. I'm not sure how to check this though. It _seemed_ to go in the correct order based on the browser cookie update and then going to the download page. But I'm not 100% on it. Is there a way to check from the downloaded file itself?

The only solution I can think of if this is an issue is to disable* the download links while the checkbox is disabled, so it's not possible to click until the attribution request has finished processing. (*disabling a link seems problematic though)

https://github.com/user-attachments/assets/b396288d-073b-4dd5-b745-503f52f23ab8

**Review by maureenlholland — APPROVED** [review]
r+wc

The only thing that gives me pause is the possible case where a user initially lands on the page and passes requirements but later returns to the page and doesn't and no longer has access to the checkbox. I think this would be very rare as the stub attribution cookies only persists for 24 hours. But it is possible if someone is travelling/on VPN/changing their cookie settings.

The other two points for consideration are:
1. Is the check for whether stub attribution exists at the right place in the meetsRequirements conditional checklist? (is there a possible issue if a user has analytics consent cookie true but stub attribution is undefined?)
2. Is there a potential race condition between a user checkbox click and download link click? 

Tested with Chrome/Safari on macOS

Tested with Browserstack
- Windows 8.1 showed expected ESR
- Windows 10/11 showed expected checkbox or no checkbox based on switch, geo, and translation availability 

UTM params were preserved when download as default didn't apply. Overridden as expected when download as default did apply.

Toggling on and off the checkboxes was reflected across all checkboxes on the page and updated URL/stub attribution as expected.

DNT on and rejected analytics cookie cases behaved as expected.

File: media/js/firefox/new/desktop/download-as-default.es6.js (hunk 4 above)
Comment (stephaniehobson) [reply media/js/firefox/new/desktop/download-as-default.es6.js:102]: I agree, the code executes in the right order to avoid the unwanted behaviour. I think we can test with a slow network connection to be extra sure.

File: media/js/firefox/new/desktop/download-as-default.es6.js (hunk 3 above)
Comment (stephaniehobson) [reply media/js/firefox/new/desktop/download-as-default.es6.js:227]: We can definitely address this in a follow up. If they return in the EU they would also have to consent before we can use stub attribution (and then I think it would re-initialize) so the only problematic case would be if they have a cookie explicitly consenting, visit in the US, and then return in the EU.

File: media/js/firefox/new/desktop/download-as-default.es6.js (hunk 4 above)
Comment (stephaniehobson) [reply media/js/firefox/new/desktop/download-as-default.es6.js:102]: Dan tested this with his crummy internet connection and it does the right thing.

**Comment by stephaniehobson** [comment]
I changed the consent cookie check so that it returns false if they're rejected but continues with the evaluation if it's set to true (so the stub attribution check will run).
